contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java

```diff
@@ -0,0 +1,928 @@
+      checkState(topics.isEmpty(), "Only topics or topicPartitions can be set, not both");
+
+      return new Read<K, V>(topics, ImmutableList.copyOf(topicPartitions), keyCoder, valueCoder,
+          timestampFn, watermarkFn, consumerFactoryFn, consumerConfig, maxNumRecords, maxReadTime);
+    }
+
+    /**
+     * Returns a new {@link Read} with {@link Coder} for key bytes.
+     * <p> Since this changes the type for key, settings that depend on the type
+     * ({@link #withTimestampFn(SerializableFunction)} and
+     *  {@link #withWatermarkFn(SerializableFunction)}) make sense only after the coders are set.
+     */
+    public <KeyT> Read<KeyT, V> withKeyCoder(Coder<KeyT> keyCoder) {
+      checkState(timestampFn == null, "Set timestampFn after setting key and value coders");
+      checkState(watermarkFn == null, "Set watermarkFn after setting key and value coders");
+      return new Read<KeyT, V>(topics, topicPartitions, keyCoder, valueCoder, null, null,
+          consumerFactoryFn, consumerConfig, maxNumRecords, maxReadTime);
+    }
+
+    /**
+     * Returns a new {@link Read} with {@link Coder} for value bytes.
+     * <p> Since this changes the type for key, settings that depend on the type
+     * ({@link #withTimestampFn(SerializableFunction)} and
+     *  {@link #withWatermarkFn(SerializableFunction)}) make sense only after the coders are set.
+     */
+    public <ValueT> Read<K, ValueT> withValueCoder(Coder<ValueT> valueCoder) {
+      checkState(timestampFn == null, "Set timestampFn after setting key and value coders");
+      checkState(watermarkFn == null, "Set watermarkFn after setting key and value coders");
+      return new Read<K, ValueT>(topics, topicPartitions, keyCoder, valueCoder, null, null,
+          consumerFactoryFn, consumerConfig, maxNumRecords, maxReadTime);
+    }
+
+    /**
+     * A function to assign a timestamp to a record. Default is processing timestamp.
+     */
+    public Read<K, V> withTimestampFn(SerializableFunction<KV<K, V>, Instant> timestampFn) {
+      checkNotNull(timestampFn);
+      return new Read<K, V>(topics, topicPartitions, keyCoder, valueCoder,
+          timestampFn, watermarkFn, consumerFactoryFn, consumerConfig, maxNumRecords, maxReadTime);
+    }
+
+    /**
+     * A function to calculate watermark after a record. Default is last record timestamp
+     * @see #withTimestampFn(SerializableFunction)
+     */
+    public Read<K, V> withWatermarkFn(SerializableFunction<KV<K, V>, Instant> watermarkFn) {
+      checkNotNull(watermarkFn);
+      return new Read<K, V>(topics, topicPartitions, keyCoder, valueCoder,
+          timestampFn, watermarkFn, consumerFactoryFn, consumerConfig, maxNumRecords, maxReadTime);
+    }
+
+    /**
+     * A factory to create Kafka {@link Consumer} from consumer configuration.
+     * Mainly used for tests. Default is {@link KafkaConsumer}.
+     */
+    public Read<K, V> withConsumerFactoryFn(
```

Comment:
if it's only used for tests, make it package-private and annotate with `@VisibleForTesting`.